src/mast/mod.rs

```diff
@@ -168,6 +171,54 @@ impl MastCtx {
     }
 }
 
+impl Symbolic {
+    /// Evaluate symbolic size to an integer.
+    pub fn eval(&self, typed_fn_env: &MonomorphizedFnEnv) -> u32 {
+        match self {
+            Symbolic::Concrete(v) => *v,
+            Symbolic::Generic(g) => typed_fn_env.get_type_info(&g.value).unwrap().value.unwrap(),
+            Symbolic::Add(a, b) => a.eval(typed_fn_env) + b.eval(typed_fn_env),
+            Symbolic::Mul(a, b) => a.eval(typed_fn_env) * b.eval(typed_fn_env),
+        }
+    }
+}
+
+impl GenericParameters {
+    /// Return all generic parameter names
+    pub fn names(&self) -> HashSet<String> {
+        self.0.keys().cloned().collect()
+    }
+
+    /// Add an unbound generic parameter
+    pub fn add(&mut self, name: String) {
+        self.0.insert(name, GenericValue::Unbound);
+    }
+
+    /// Bind a generic parameter to a value
+    pub fn bind(&mut self, name: String, value: u32, span: Span) -> Result<()> {
+        let existing = self.0.get(&name);
+        match existing {
+            Some(GenericValue::Bound(v)) => {
+                if *v == value {
+                    return Ok(());
+                }
+
+                Err(Error::new(
+                    "mast",
+                    ErrorKind::ConflictGenericValue(name, *v, value),
+                    span,
+                ))
+            }
+            Some(GenericValue::Unbound) => Ok(()),
```

Comment:
isn't this where you'd want to set the value?

Reply:
oh right!